karapace/backup/topic_configurations.py

```diff
@@ -0,0 +1,73 @@
+"""
+Copyright (c) 2023 Aiven Ltd
+See LICENSE for details
+"""
+from enum import Enum
+from kafka import KafkaAdminClient
+from kafka.admin import ConfigResource, ConfigResourceType
+from kafka.errors import for_code
+from kafka.protocol.admin import DescribeConfigsRequest
+from typing import Dict, Iterable, Optional
+
+
+class ConfigSource(int, Enum):
+    UNKNOWN = 0
+    TOPIC_CONFIG = 1
+    DYNAMIC_BROKER_CONFIG = 2
+    DYNAMIC_DEFAULT_BROKER_CONFIG = 3
+    STATIC_BROKER_CONFIG = 4
+    DEFAULT_CONFIG = 5
+    DYNAMIC_BROKER_LOGGER_CONFIG = 6
+
+
+DEFAULT_CONFIGS = [
+    "cleanup.policy",
+    "min.insync.replicas",
+    "retention.bytes",
+    "retention.ms",
+]
+
+
+def get_topic_configurations(
+    admin: KafkaAdminClient,
+    topic_name: str,
+    config_source_filter: Optional[Iterable[ConfigSource]] = None,
+) -> Dict[str, str]:
+    """Get configurations of the specified topic. The following configurations will be retrieved by default:
+        - `cleanup.policy`
+        - `min.insync.replicas`
+        - `retention.bytes`
+        - `retention.ms`
+
+    :param admin: Kafka admin client
+    :param topic_name: get configurations for this topic
+    :param config_source_filter: if specified, returns all the configurations that match the sources specified,
+        plus the default configurations. If None, returns only the default configurations.
+    """
+    config_version = admin._matching_api_version(DescribeConfigsRequest)  # pylint: disable=protected-access
+    req_cfgs = [ConfigResource(ConfigResourceType.TOPIC, topic_name)]
+    cfgs = admin.describe_configs(req_cfgs)
+    assert len(cfgs) == 1
+    assert len(cfgs[0].resources) == 1
+    err, _, _, _, config_values = cfgs[0].resources[0]
+    if err != 0:
+        raise for_code(err)
+    topic_config = {}
+    for cv in config_values:
+        if config_version == 0:
+            name, val, _, _, _ = cv
+            # DescribeConfigs Response v0 does not expose config source: retrieve just a subset of basic configs
+            if name in DEFAULT_CONFIGS:
+                topic_config[name] = val
+        else:
+            name, val, _, config_source, _, _ = cv
+            if name in DEFAULT_CONFIGS or (config_source_filter is not None and config_source in config_source_filter):
+                topic_config[name] = val
```

Comment:
question: Do we have test coverage of both these versions of `DescribeConfig`?

Reply:
I added some integration tests for `get_topic_configurations`. Unfortunately they don't cover DescribConfig response v0, because it's only returned by very old brokers

Reply:
@giuseppelillo Do we need to support it? If all versions of it are already EOL, I'd argue for not including that support.